### PR TITLE
Allowing bindmounts to mount anywhere

### PR DESCRIPTION
**Pull request by DunderRoffe**

### CODE REVIEW

File: libsoftwarecontainer/softwarecontainer.conf.in [hunk 1]
```diff
@@ -9,6 +9,9 @@ lxc.pts = 1
 #
 lxc.haltsignal = SIGKILL
 
+lxc.loglevel = debug
+lxc.logfile = /tmp/sc_lxc.log
```
Comment (tobsan) [inline libsoftwarecontainer/softwarecontainer.conf.in]: Please remove this.

File: libsoftwarecontainer/src/container.cpp [hunk 2]
```diff
@@ -574,62 +574,92 @@ ReturnCode Container::destroy(unsigned int timeout)
...
     }
 
-    result = gatewaysDirInContainer() + "/" + pathInContainer;
+    std::string tempDirInContainer = gatewaysDirInContainer() + "/" + std::string(basename(pathInContainer.c_str()));
+
+    // Move the mount in the container
+    if (tempDirInContainer.compare(pathInContainer) != 0) {
+        pid_t pid = INVALID_PID;
+        ReturnCode mountMoveRes = executeInContainer([tempDirInContainer, pathInContainer] () {
+            unsigned long flags = MS_MOVE;
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp:639]: Are you sure you don't want to keep the flags used before and just or with MS_MOVE?

File: libsoftwarecontainer/src/container.cpp [hunk 3]
```diff
@@ -574,62 +574,92 @@ ReturnCode Container::destroy(unsigned int timeout)
...
+            if (isDirectory(tempDirInContainer)) {
+                mkdir(pathInContainer.c_str(), 1);
```
Comment (tobsan) [inline libsoftwarecontainer/src/container.cpp]: What does the 1 mean here? There should be predefined constants for this.

File: libsoftwarecontainer/src/gateway/pulsegateway.cpp [hunk 4]
```diff
@@ -42,27 +42,31 @@ ReturnCode PulseGateway::readConfigElement(const json_t *element)
...
+    std::string pathInContainer = "/gateways/" + std::string(SOCKET_FILE_NAME);
+    ReturnCode result = getContainer()->bindMountFileInContainer(std::string(dir), pathInContainer, false);
+
+    if (isError(result)) {
+        log_error() << "Could not bind mount pulseaudio socket in container";
+        return false;
+    }
+
+    std::string unixPath = "unix:" + pathInContainer;
+    setEnvironmentVariable(PULSE_AUDIO_SERVER_ENVIRONMENT_VARIABLE_NAME, unixPath );
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/pulsegateway.cpp]: What's with the trailing space here?

**Review by tobsan — CHANGES_REQUESTED** [review]
I would like to see additional tests, like what happens if you try to mount "/" or something where things are already mounted to. Apart from that, this looks good.

File: libsoftwarecontainer/src/container.cpp (hunk 2 above)
Comment (DunderRoffe) [reply libsoftwarecontainer/src/container.cpp:639]: when using Move all other flags are ignored anyway so it does not matter

File: libsoftwarecontainer/src/container.cpp (hunk 3 above)
Comment (DunderRoffe) [reply libsoftwarecontainer/src/container.cpp]: sure

File: libsoftwarecontainer/src/gateway/pulsegateway.cpp (hunk 4 above)
Comment (DunderRoffe) [reply libsoftwarecontainer/src/gateway/pulsegateway.cpp]: I fix

File: libsoftwarecontainer/src/container.cpp [hunk 5]
```diff
@@ -766,6 +796,11 @@ std::string Container::gatewaysDirInContainer() const
...
 }
 
+std::string Container::containerRoot() const
+{
+    return m_containerRoot + "/" + id();
+}
+
 std::string Container::gatewaysDir() const
 {
     return m_containerRoot + "/" + id() + GATEWAYS_PATH;
```
Comment (kursatkobya) [inline libsoftwarecontainer/src/container.cpp:801]: maybe it could be good idea to use containerRoot() here instead of `m_containerRoot + "/" + id()`

Comment (DunderRoffe) [reply libsoftwarecontainer/src/container.cpp:801]: actually I don't think containerRoot is used, maybe it should be removed

**Review by tobsan — APPROVED** [review]

**Pull request merged**